crates/core/component/shielded-pool/src/component/rpc/bank_query.rs

```diff
@@ -0,0 +1,198 @@
+            .into_iter()
+            .collect::<BTreeMap<_, _>>();
+
+        let s = snapshot.prefix(ibc_state_key::ics20_value_balance::prefix());
+        let ibc_amounts = s
+            .filter_map(move |i: anyhow::Result<(String, Amount)>| async move {
+                if i.is_err() {
+                    return Some(Err(i.context("bad amount in state").err().unwrap()));
+                }
+                let (key, amount) = i.expect("should not be an error");
+
+                // Extract the asset ID from the key
+                let asset_id = key.split('/').last();
+                if asset_id.is_none() {
+                    return Some(Err(asset_id
+                        .context("bad IBC ics20 value balance key in state")
+                        .err()
+                        .unwrap()));
+                }
+                let asset_id = asset_id.expect("should not be an error");
+
+                // Parse the asset ID
+                let asset_id = asset_id.parse::<asset::Id>();
+                if asset_id.is_err() {
+                    return Some(Err(asset_id
+                        .context("invalid IBC ics20 value balance asset ID in state")
+                        .err()
+                        .unwrap()));
+                }
+                let asset_id = asset_id.expect("should not be an error");
+
+                Some(Ok((asset_id, amount)))
+            })
+            .collect::<Vec<_>>()
+            .await
+            .into_iter()
+            .collect::<anyhow::Result<Vec<_>>>()
+            .map_err(|e| tonic::Status::internal(e.to_string()))?;
+
+        // Fetch the denoms associated with the IBC asset IDs
+        for (asset_id, amount) in ibc_amounts {
+            let denom_metadata = snapshot.denom_metadata_by_asset(&asset_id).await;
+            if denom_metadata.is_none() {
+                return Err(tonic::Status::internal(
+                    "bad IBC ics20 value balance key in state".to_string(),
+                ));
+            }
+            let denom_metadata = denom_metadata.expect("should not be an error");
+
+            total_supply.insert(denom_metadata, amount);
+        }
+
+        Ok(tonic::Response::new(QueryTotalSupplyResponse {
+            // Pagination disabled for now
+            pagination: None,
+            supply: total_supply
+                .into_iter()
+                .map(
+                    |(denom_metadata, amount)| ibc_proto::cosmos::base::v1beta1::Coin {
+                        denom: denom_metadata.to_string(),
+                        amount: amount.to_string(),
+                    },
+                )
+                .collect::<Vec<ibc_proto::cosmos::base::v1beta1::Coin>>(),
+        }))
+    }
+
+    async fn params(
+        &self,
+        _: tonic::Request<QueryParamsRequest>,
+    ) -> std::result::Result<tonic::Response<QueryParamsResponse>, tonic::Status> {
+        unimplemented!()
```

Comment:
```suggestion
        Err(Status::unimplemented("params endpoint is not implemented"))
```

can we return an unimplemented `Status` rather than calling `unimplemented!()` here? this way these endpoints won't allow a malicious caller to cause a node to panic.

Reply:
@cratelyn good catch, I see some other uses of `unimplemented!()` elsewhere in the codebase, we should probably audit those and make sure they're not triggerable in the nodes

Reply:
agreed! with #4446 filed and 8a0c1d9 in hand, i'll mark this as resolved. thank you!